### DISCUSSION TITLE
Need help handling "Expected corresponding JSX closing tag for <blah>"

### DISCUSSION THREAD

**Discussion by oawad79** [facebook/docusaurus#9370] [General]
I am working on using Docusaurus to create a central location for markdowns coming from many sources, I have written a GitLab plugin to load data from GitLab by querying the GitLab API using Axios and writing them under the "docs" directory under Docusaurus, so we are talking about thousands of markdowns, I keep getting issues related to broken markdown links which I set to ignore in the docusaurus.config.js file and that worked fine, but now I am getting lots of issues related to missing JSX tags, I do understand that I can do some post-processing to either escape the tags or close them if possible ... is there a way to make Docusaurus be more lenient toward this as this is a lot of post-processing and I keep finding more and more corner cases...?

**Comment by slorber** [comment]
Hi

It's impossible to help without a repro

> I keep getting issues related to broken markdown links which I set to ignore in the docusaurus.config.js file

If links are broken you'd rather fix them one by one, or your users will navigate to 404 pages. 



> now I am getting lots of issues related to missing JSX tags

Make sure to understand how Docusaurus and MDX works.
https://docusaurus.io/docs/markdown-features

Docusaurus is using MDX and not CommonMark. There are differences to take into consideration: you are not using embedded HTML anymore but JSX.

Docusaurus v2 always uses MDX, but with Docusaurus v3 (in beta, release soon) it's possible to use CommonMark (with some limitations that we'll fix, see https://github.com/facebook/docusaurus/issues/9092 for details)

**Reply by oawad79** [reply]
Hello,

So, I am loading thousands of markdown documents from many sources, not possible for me to fix them manually, it is okay if they point to a broken location as many of them are old documents that point to old sites that don't exist anymore, I still need the document to load so setting ignore onBrokenMarkdownLinks is good enough for me.

I do understand that Docusaurus uses JSX, I tried to load the files as MD files instead of MDX to avoid JSX but doesn't seem to resolve the issue since, as you stated, Docusaurus uses JSX all the time and I guess that is my issue.

I just tried Docusaurus 3.00-beta to see if it will help but getting the same issues on missing closed tags... not sure if it is more forgiving, so for example
```markdown
# <ins> Table Of Contents
### [Summary](#summaries)
### [Technologies](#tech)
### [Useful Links/Info](#useful-links)
### [Integrations](#integration)
### [Example Requests](#example-request)
### [Special Cases](#special-case)
### [Swagger](#swag)
### [Cache](#cache)
### [Error Codes by Endpoint](#error-codes)
### [Creating API Credentials for Third Parties](#third-party-credentials)

# <ins>Summary</ins> <a name="summaries"/>

<App>
```
fails on two locations, `<ins>` and `<App>` does not having closing tags
 

So, given that, I did manage to write a plugin to rewrite the broken images and some of the Markdowns using regex, I am looking into AST and unifedjs to try to remove unclosed JSX tags, is there something else I can do to make this easier? 

is there a way to make Docusaurus load markdowns only without trying to load them as MDX? it still can do MDX but it will load .MD files as markdowns and not MDX?

**Reply by oawad79** [reply]
As mentioned in your 

https://github.com/facebook/docusaurus/issues/9092

 This seems to work so far

![image](https://github.com/facebook/docusaurus/assets/7880654/161f02fd-2246-4833-a405-75c8175df986)

the files I am testing with loads without issues

**Reply by slorber** [reply]
Yes, you should try Docusaurus v3 with `siteConfig.markdown.format: "detect"` and see if it works for you. There will certainly be some limitations for now that we'll fix progressively in v3.x, but it should work in most cases for now.

This mode is important for us in particular for such use cases where a large doc want to adopt Docusaurus, without needing to convert the existing docs.

A remark plugin can't handle that because it only works once the AST is available. Here the problem is that your input doc does not compile with MDX so it is not even transformed to AST.

We also added another workaround option in v3:

```js
const config = {
  markdown: {
    format: 'detect', 
    // Gives you opportunity to apply your own string processing before the MDX compilation
    // Usage is NOT recommended, use this as a last resort escape hatch or temporarily)
    preprocessor: ({ filePath, fileContent }) => {
      return fileContent;
    }
  }
};
```

**Reply by oawad79** [reply]
Thank you, this explains why the `unist-util-visit` does not work anymore...or am I missing something here? the work around you provided worked so far. I am using it to remove empty markdown links like `[something]()` and worked 

However, I am getting this when I do npm install

```
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR!
npm ERR! While resolving: @cmfcmf/docusaurus-search-local@1.1.0
npm ERR! Found: @docusaurus/core@3.0.0-beta.0
npm ERR! node_modules/@docusaurus/core
npm ERR!   @docusaurus/core@"3.0.0-beta.0" from the root project
npm ERR!   @docusaurus/core@"3.0.0-beta.0" from @docusaurus/plugin-content-blog@3.0.0-beta.0
npm ERR!   node_modules/@docusaurus/plugin-content-blog
npm ERR!     @docusaurus/plugin-content-blog@"3.0.0-beta.0" from @docusaurus/preset-classic@3.0.0-beta.0
npm ERR!     node_modules/@docusaurus/preset-classic
npm ERR!       @docusaurus/preset-classic@"3.0.0-beta.0" from the root project
npm ERR!     @docusaurus/plugin-content-blog@"3.0.0-beta.0" from @docusaurus/theme-classic@3.0.0-beta.0
npm ERR!     node_modules/@docusaurus/theme-classic
npm ERR!       @docusaurus/theme-classic@"3.0.0-beta.0" from @docusaurus/preset-classic@3.0.0-beta.0
npm ERR!       node_modules/@docusaurus/preset-classic
npm ERR!         @docusaurus/preset-classic@"3.0.0-beta.0" from the root project
npm ERR!     1 more (@docusaurus/theme-common)
npm ERR!   11 more (@docusaurus/plugin-content-docs, ...)
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! peer @docusaurus/core@"^2.0.0" from @cmfcmf/docusaurus-search-local@1.1.0
npm ERR! node_modules/@cmfcmf/docusaurus-search-local
npm ERR!   @cmfcmf/docusaurus-search-local@"^1.1.0" from the root project
npm ERR!
npm ERR! Conflicting peer dependency: @docusaurus/core@2.4.3
npm ERR! node_modules/@docusaurus/core
npm ERR!   peer @docusaurus/core@"^2.0.0" from @cmfcmf/docusaurus-search-local@1.1.0
npm ERR!   node_modules/@cmfcmf/docusaurus-search-local
npm ERR!     @cmfcmf/docusaurus-search-local@"^1.1.0" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR!
npm ERR!
```
when I add the `--legacy-peer-deps`, it works fine, is this because other plugins are not yet upgraded to work with V3? what should I add to the "peerDependencies" section in the package.json?

**Reply by slorber** [reply]
I don't know what you mean by "unist-util-visit does not work anymore", without more details it's impossible to help. My guess is that you possibly get an error because newer versions of this packages are now ESM-only and not CJS.

The search plugin you mention doesn't seem to support Docusaurus v3 yet: https://github.com/cmfcmf/docusaurus-search-local/issues/194

**Reply by oawad79** [reply]
that is correct, I got that error but downgrading to an older version worked

**Reply by oawad79** [reply]
I have managed to get search local plugin to work with V3 by updating the dependencies for the plugin to v3 beta and disabling the sidebar versioning in the plugin since it keeps complaining that Docusarus 3 no longer supports that, however, I am still getting issues with peer dependencies

**Reply by slorber** [reply]
If you have an issue you need to invest your own time to create a repro and save me time, otherwise I won't be able to help.

**Comment by oawad79** [comment]
wonder what changed that made the search local not work for the v3
